File: generate-feed.mjs
#!/usr/bin/env node
/**
 * Generates feed.json for a consciousness studies feed.
 * Sources: arXiv API (official Atom feed format).
 *
 * No external dependencies required.
 */

const fs = require("fs/promises");

const OUT_FILE = "feed.json";
const MAX_ITEMS = 24;

// Tune these queries as needed.
const SEARCHES = [
  {
    source: "arXiv · consciousness",
    query: 'all:consciousness OR all:"neural correlates of consciousness" OR all:"global workspace"'
  },
  {
    source: "arXiv · anesthesia",
    query: 'all:anesthesia OR all:anaesthesia OR all:"loss of consciousness"'
  },
  {
    source: "arXiv · meditation",
    query: 'all:meditation OR all:awareness OR all:attention OR all:"altered states"'
  },
  {
    source: "arXiv · philosophy of mind",
    query: 'all:"integrated information" OR all:"phi" OR all:"philosophy of mind" OR all:"machine consciousness"'
  }
];

function escapeXml(s = "") {
  return String(s)
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;")
    .replace(/'/g, "&apos;");
}

function stripTags(s = "") {
  return String(s).replace(/<[^>]+>/g, " ").replace(/\s+/g, " ").trim();
}

function htmlToText(s = "") {
  return stripTags(s)
    .replace(/&amp;/g, "&")
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&quot;/g, '"')
    .replace(/&apos;/g, "'");
}

function truncate(s, n) {
  const text = String(s || "").trim();
  if (text.length <= n) return text;
  return text.slice(0, Math.max(0, n - 1)).trimEnd() + "…";
}

function parseArxivFeed(xml, sourceLabel) {
  const entries = [];
  const entryMatches = xml.match(/<entry>[\s\S]*?<\/entry>/g) || [];

  for (const entry of entryMatches) {
    const get = (tag) => {
      const m = entry.match(new RegExp(`<${tag}>([\\s\\S]*?)<\\/${tag}>`));
      return m ? htmlToText(m[1]) : "";
    };

    const title = get("title").replace(/\s+/g, " ").trim();
    const summary = get("summary").replace(/\s+/g, " ").trim();
    const published = get("published").slice(0, 10);
    const updated = get("updated").slice(0, 10);

    const idMatch = entry.match(/<id>([\s\S]*?)<\/id>/);
    const linkMatch = entry.match(/<link[^>]*rel="alternate"[^>]*href="([^"]+)"/) || entry.match(/<id>([\s\S]*?)<\/id>/);

    const authors = [...entry.matchAll(/<author>[\s\S]*?<name>([\s\S]*?)<\/name>[\s\S]*?<\/author>/g)]
      .map(m => htmlToText(m[1]))
      .filter(Boolean)
      .slice(0, 3)
      .join(", ");

    entries.push({
      title,
      summary: truncate(summary, 180),
      source: sourceLabel,
      date: published || updated || new Date().toISOString().slice(0, 10),
      link: linkMatch ? htmlToText(linkMatch[1] || linkMatch[0]) : (idMatch ? htmlToText(idMatch[1]) : "https://arxiv.org/"),
      authors
    });
  }

  return entries;
}

async function fetchWithTimeout(url, timeoutMs = 20000) {
  const controller = new AbortController();
  const t = setTimeout(() => controller.abort(), timeoutMs);
  try {
    const res = await fetch(url, {
      signal: controller.signal,
      headers: {
        "User-Agent": "Mozilla/5.0 (compatible; ConsciousnessFeed/1.0; +https://github.com/tasselrex/AIResearchFeed)"
      }
    });
    if (!res.ok) throw new Error(`HTTP ${res.status} ${res.statusText}`);
    return await res.text();
  } finally {
    clearTimeout(t);
  }
}

async function gather() {
  const all = [];
  for (const s of SEARCHES) {
    const url = new URL("https://export.arxiv.org/api/query");
    url.searchParams.set("search_query", s.query);
    url.searchParams.set("start", "0");
    url.searchParams.set("max_results", "8");
    url.searchParams.set("sortBy", "submittedDate");
    url.searchParams.set("sortOrder", "descending");

    try {
      const xml = await fetchWithTimeout(url.toString());
      const items = parseArxivFeed(xml, s.source);
      all.push(...items);
    } catch (err) {
      console.error(`Failed to fetch ${s.source}:`, err.message);
    }
  }

  // Deduplicate by title and keep the newest, most relevant ones.
  const seen = new Set();
  const deduped = [];
  for (const item of all) {
    const key = item.title.toLowerCase();
    if (seen.has(key)) continue;
    seen.add(key);
    deduped.push(item);
  }

  // Prefer newer items; keep stable order within same date.
  deduped.sort((a, b) => String(b.date).localeCompare(String(a.date)));

  const payload = {
    updatedAt: new Date().toISOString(),
    items: deduped.slice(0, MAX_ITEMS)
  };

  await fs.writeFile(OUT_FILE, JSON.stringify(payload, null, 2) + "\n", "utf8");
  console.log(`Wrote ${OUT_FILE} with ${payload.items.length} items`);
}

gather().catch(err => {
  console.error(err);
  process.exit(1);
});

File: generate-feed.py
#!/usr/bin/env python3
from __future__ import annotations

import json
import re
import urllib.parse
import urllib.request
import xml.etree.ElementTree as ET
from datetime import datetime, timezone
from pathlib import Path

OUT = Path("feed.json")
MAX_ITEMS = 24

QUERIES = [
    ("arXiv · consciousness", 'all:consciousness OR all:"global workspace" OR all:"neural correlates of consciousness"'),
    ("arXiv · anesthesia", 'all:anesthesia OR all:anaesthesia OR all:"loss of consciousness"'),
    ("arXiv · philosophy of mind", 'all:"philosophy of mind" OR all:"integrated information" OR all:"machine consciousness"'),
    ("arXiv · meditation", 'all:meditation OR all:awareness OR all:attention OR all:"altered states"'),
]

HEADERS = {
    "User-Agent": "Mozilla/5.0 (compatible; ConsciousnessFeed/1.0; +https://github.com/tasselrex/AIResearchFeed)"
}

def clean(s: str | None) -> str:
    if not s:
        return ""
    s = re.sub(r"\s+", " ", s)
    return s.strip()

def strip_html(s: str | None) -> str:
    if not s:
        return ""
    s = re.sub(r"<[^>]+>", " ", s)
    s = clean(s)
    return (
        s.replace("&amp;", "&")
         .replace("&lt;", "<")
         .replace("&gt;", ">")
         .replace("&quot;", '"')
         .replace("&apos;", "'")
    )

def truncate(s: str, n: int = 180) -> str:
    s = clean(s)
    return s if len(s) <= n else s[: n - 1].rstrip() + "…"

def fetch(url: str, timeout: int = 25) -> str:
    req = urllib.request.Request(url, headers=HEADERS)
    with urllib.request.urlopen(req, timeout=timeout) as resp:
        return resp.read().decode("utf-8", errors="replace")

def parse_atom(xml_text: str, source_label: str):
    ns = {"a": "http://www.w3.org/2005/Atom"}
    root = ET.fromstring(xml_text)
    items = []
    for entry in root.findall("a:entry", ns):
        title = clean(entry.findtext("a:title", default="", namespaces=ns))
        summary = truncate(strip_html(entry.findtext("a:summary", default="", namespaces=ns)), 180)
        published = clean(entry.findtext("a:published", default="", namespaces=ns))[:10]
        updated = clean(entry.findtext("a:updated", default="", namespaces=ns))[:10]

        link = ""
        for link_el in entry.findall("a:link", ns):
            if link_el.attrib.get("rel", "alternate") == "alternate" and link_el.attrib.get("href"):
                link = link_el.attrib["href"]
                break
        if not link:
            id_text = clean(entry.findtext("a:id", default="", namespaces=ns))
            link = id_text

        authors = [
            clean(author.findtext("a:name", default="", namespaces=ns))
            for author in entry.findall("a:author", ns)
        ]
        authors = ", ".join([a for a in authors if a][:3])

        items.append({
            "title": title,
            "summary": summary,
            "source": source_label,
            "date": published or updated or datetime.now(timezone.utc).date().isoformat(),
            "link": link,
            "authors": authors,
        })
    return items

def arxiv_url(query: str) -> str:
    base = "https://export.arxiv.org/api/query"
    params = {
        "search_query": query,
        "start": "0",
        "max_results": "10",
        "sortBy": "submittedDate",
        "sortOrder": "descending",
    }
    return base + "?" + urllib.parse.urlencode(params, safe=':" OR')

def main():
    all_items = []
    errors = []
    for label, query in QUERIES:
        try:
            xml_text = fetch(arxiv_url(query))
            items = parse_atom(xml_text, label)
            all_items.extend(items)
        except Exception as e:
            errors.append(f"{label}: {e}")

    dedup = []
    seen = set()
    for item in all_items:
        key = item["title"].lower()
        if key in seen:
            continue
        seen.add(key)
        dedup.append(item)

    dedup.sort(key=lambda x: x["date"], reverse=True)

    payload = {
        "updatedAt": datetime.now(timezone.utc).isoformat(),
        "items": dedup[:MAX_ITEMS],
        "errors": errors,
    }
    OUT.write_text(json.dumps(payload, indent=2) + "\n", encoding="utf-8")
    print(f"Wrote {OUT} with {len(payload['items'])} items")

if __name__ == "__main__":
    main()
